test/e2e/gameserver_test.go

```diff
@@ -213,6 +213,113 @@ func TestGameServerUnhealthyAfterDeletingPod(t *testing.T) {
 	assert.NoError(t, err)
 }
 
+func TestGameServerRestartBeforeReadyCrash(t *testing.T) {
+	t.Parallel()
+	logger := logrus.WithField("test", t.Name())
+
+	gs := defaultGameServer(defaultNs)
+	// give some buffer with gameservers crashing and coming back
+	gs.Spec.Health.PeriodSeconds = 60 * 60
+	gs.Spec.Template.Spec.Containers[0].Env = append(gs.Spec.Template.Spec.Containers[0].Env, corev1.EnvVar{Name: "READY", Value: "FALSE"})
+	gsClient := framework.AgonesClient.AgonesV1().GameServers(defaultNs)
+	newGs, err := gsClient.Create(gs)
+	if !assert.NoError(t, err) {
+		assert.Fail(t, "could not create the gameserver")
+	}
+	defer gsClient.Delete(newGs.ObjectMeta.Name, nil) // nolint: errcheck
+
+	logger.Info("Waiting for us to have an address to send things to")
+	newGs, err = framework.WaitForGameServerState(newGs, agonesv1.GameServerStateScheduled, time.Minute)
+	assert.NoError(t, err)
+
+	logger.WithField("gs", newGs.ObjectMeta.Name).Info("GameServer created")
+
+	address := fmt.Sprintf("%s:%d", newGs.Status.Address, newGs.Status.Ports[0].Port)
+	logger.WithField("address", address).Info("Dialing UDP message to address")
+
+	logger.Info("crashing, and waiting to see restart")
+	err = wait.PollImmediate(3*time.Second, 3*time.Minute, func() (bool, error) {
+		pod, err := framework.KubeClient.CoreV1().Pods(defaultNs).Get(newGs.ObjectMeta.Name, metav1.GetOptions{})
+		if err != nil {
+			logger.WithError(err).Warn("could not get pod for gameserver")
+			return true, err
+		}
+
+		for _, c := range pod.Status.ContainerStatuses {
+			if c.Name == newGs.Spec.Container && c.RestartCount >= 1 {
+				logger.Info("successfully crashed. Moving on!")
+				return true, nil
+			}
+		}
+
+		// create a connection each time, as weird stuff happens if the receiver isn't up and running.
+		conn, err := net.Dial("udp", address)
+		assert.NoError(t, err)
+		defer conn.Close() // nolint: errcheck
+		// doing this last, so that there is a short delay between the message being sent, and the check.
+		// also gives the process a chance to come back up
+		_, err = conn.Write([]byte("CRASH"))
+		logger.WithError(err).Info("no crash, trying again")
+		return false, nil
```

Comment:
should this return `false, err` instead of nil? Or will returning an error short circuit the polling function? It seems weird that the error writing the udp packet is logged but otherwise ignored.

Reply:
You are indeed correct - returning the error will short circuit the poll. It may return an error as the receiver may not be up and running yet, so it's useful for debugging purposes, but otherwise it is expected.